app/src/main/java/br/com/dcarv/arcmovies/presentation/view/MainActivity.kt

```diff
@@ -0,0 +1,293 @@
+package br.com.dcarv.arcmovies.presentation.view
+
+import android.content.Intent
+import android.os.Bundle
+import android.os.Handler
+import android.support.v4.app.ActivityOptionsCompat
+import android.support.v4.view.ViewCompat
+import android.support.v7.app.AppCompatActivity
+import android.support.v7.widget.DividerItemDecoration
+import android.support.v7.widget.LinearLayoutManager
+import android.support.v7.widget.RecyclerView
+import android.support.v7.widget.SearchView
+import android.util.Log
+import android.view.Menu
+import android.view.MenuItem
+import br.com.dcarv.arcmovies.*
+import br.com.dcarv.arcmovies.data.prefs.SharedPreferencesProvider
+import br.com.dcarv.arcmovies.data.tmdb.*
+import br.com.dcarv.arcmovies.data.tmdb.model.TMDBCountryMapper
+import br.com.dcarv.arcmovies.data.tmdb.model.TMDBGenreMapper
+import br.com.dcarv.arcmovies.data.tmdb.model.TMDBLanguageMapper
+import br.com.dcarv.arcmovies.data.tmdb.model.TMDBUpcomingMoviesListMapper
+import br.com.dcarv.arcmovies.domain.abstraction.*
+import br.com.dcarv.arcmovies.domain.model.Country
+import br.com.dcarv.arcmovies.domain.model.Language
+import br.com.dcarv.arcmovies.domain.model.UpcomingMoviesList
+import br.com.dcarv.arcmovies.presentation.presenter.IMainPresenter
+import br.com.dcarv.arcmovies.presentation.presenter.MainPresenter
+import com.jakewharton.retrofit2.adapter.rxjava2.RxJava2CallAdapterFactory
+import kotlinx.android.synthetic.main.activity_main.*
+import kotlinx.android.synthetic.main.content_main.*
+import okhttp3.OkHttpClient
+import okhttp3.logging.HttpLoggingInterceptor
+import retrofit2.Retrofit
+import retrofit2.converter.gson.GsonConverterFactory
+
+/**
+ * An implementation of IMainActivity for Android.
+ * Shows the list of upcoming movies
+ *
+ * @author Danilo Carvalho
+ */
+class MainActivity : AppCompatActivity(), IMainView,
+    SearchView.OnQueryTextListener, RegionSettingsPickerDialogFragment.Listener {
+    // Presenter
+    private lateinit var presenter: IMainPresenter
+
+    // Network Dependencies
+    private lateinit var service: TMDBService
+    private lateinit var moviesProvider: IUpcomingMoviesListProvider
+    private lateinit var genreProvider: IGenreProvider
+    private lateinit var countriesProvider: ICountriesListProvider
+    private lateinit var languagesProvider: ILanguagesListProvider
+    private lateinit var transtationsProvider: ITranslationsProvider
+    private lateinit var preferencesProvider: IPreferencesProvider
+
+    // Movies Recycler Adapter
+    lateinit var adapter: MoviesAdapter
+
+    private var searchView: SearchView? = null
+
+    override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+        setContentView(R.layout.activity_main)
+        setSupportActionBar(toolbar)
+
+        // TODO: use Dependency Injection to inject these dependencies instead
```

Comment:
Do you see any way you could improve what you currently have without the use of DI?

Reply:
Yes. I could use Factories or Factory Methods to create these classes. I meant to use Dagger2 to inject the dependencies, but I've only ever played around with it, never really implemented it on a production app, so I decided not to use it in case I introduced bugs I was not aware of.